core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -742,13 +742,13 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
         private final List<ShardRouting> shards;
         private final DiscoveryNodes nodes;
         private final boolean executeOnReplica;
-        private final String indexUUID;
         private final AtomicBoolean finished = new AtomicBoolean();
         private final AtomicInteger success = new AtomicInteger(1); // We already wrote into the primary shard
         private final ConcurrentMap<String, Throwable> shardReplicaFailures = ConcurrentCollections.newConcurrentMap();
         private final AtomicInteger pending;
         private final int totalShards;
         private final Releasable indexShardReference;
+        private final ShardRouting primaryShard;
```

Comment:
now that #15900 is in, we can remove this and get the shardrouting of the primary from the indexShardReference.